File: go/algorithms/bogo.go
package algorithms

import (
    "errors"
    "math/rand"
    "time"
)

// BogoSort shuffles until sorted or maxShuffles reached; returns error if not sorted.
func BogoSort(arr []int, maxShuffles int) error {
    if len(arr) <= 1 {
        return nil
    }
    rand.Seed(time.Now().UnixNano())
    for s := 0; s < maxShuffles; s++ {
        if isSorted(arr) {
            return nil
        }
        rand.Shuffle(len(arr), func(i, j int) { arr[i], arr[j] = arr[j], arr[i] })
    }
    if !isSorted(arr) {
        return errors.New("bogo sort exceeded maxShuffles")
    }
    return nil
}

func isSorted(arr []int) bool {
    for i := 1; i < len(arr); i++ {
        if arr[i] < arr[i-1] {
            return false
        }
    }
    return true
}

File: rust/src/bogo.rs
use rand::seq::SliceRandom;
use rand::thread_rng;

pub fn bogo_sort(arr: &mut [i32], max_shuffles: usize) -> Result<(), &'static str> {
    let mut rng = thread_rng();
    for _ in 0..max_shuffles {
        if is_sorted(arr) {
            return Ok(());
        }
        arr.shuffle(&mut rng);
    }
    if is_sorted(arr) {
        Ok(())
    } else {
        Err("bogo sort exceeded max_shuffles")
    }
}

fn is_sorted(arr: &[i32]) -> bool {
    arr.windows(2).all(|w| w[0] <= w[1])
}
